### PR TITLE
[Validator] Allow to define a reusable set of constraints

### LINKED ISSUE #13075: [Validator] Allow to define a reusable set of constraints
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#34334
| PR author(s) | @ogizanagi
| Merged in    | 5.1-dev

### PR DESCRIPTION
**Pull request by ogizanagi**
closes #13075

### CODE REVIEW

File: reference/constraints/Compound.rst [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+        // src/Validator/Constraints/MatchesPasswordRequirements.php
+        namespace App\Validator\Constraints;
+
+        use Symfony\Component\Validator\Compound;
+
+        /**
+         * @Annotation
+         */
+        class MatchesPasswordRequirements extends Compound
```
Comment (HeahDude) [inline reference/constraints/Compound.rst]: `PasswordRequirementsContraint`?

File: reference/constraints/Compound.rst [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        {
+            protected function getConstraints(array $options): array
+            {
+                return [
+                    new NotBlank(),
+                    new Type('string'),
+                    new Length(['min' => 12]),
+                    new NotCompromisedPassword(),
```
Comment (HeahDude) [inline reference/constraints/Compound.rst]: All those constraints needs a use on that file, I suggest to use an alias as usual, then: 
```php
use Symfony\Component\Validator\Constraints as Assert;
//...
return [
    new Assert\NotBlank(),
    // ...
```

File: reference/constraints/Compound.rst [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        }
+
+You can now use it anywhere you need it:
+
+.. configuration-block::
+
+    .. code-block:: php-annotations
+
+        // src/Entity/User.php
+        namespace App\DTO;
```
Comment (HeahDude) [inline reference/constraints/Compound.rst]: What about `App\User` instead?

File: reference/constraints/Compound.rst [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+        use App\Validator\Constraints\MatchesPasswordRequirements;
```
Comment (HeahDude) [inline reference/constraints/Compound.rst]: I would do the following consistently with my comment above:
```php
use App\Validator\Constraints\PasswordRequirementContraint as AssertPasswordRequirements;
```

File: reference/constraints/Compound.rst [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+        class RegisterUser
+        {
+            /**
+             * @MatchesPasswordRequirements()
```
Comment (HeahDude) [inline reference/constraints/Compound.rst]: And here `@AssertPasswordRequirements`

File: validation/custom_constraint.rst [hunk 6]
```diff
@@ -181,6 +181,12 @@ then your validator is already registered as a service and :doc:`tagged </servic
 with the necessary ``validator.constraint_validator``. This means you can
 :ref:`inject services or configuration <services-constructor-injection>` like any other service.
 
+Create a reusable set of constraints
```
Comment (HeahDude) [inline validation/custom_constraint.rst]: We need a title case here

File: validation/custom_constraint.rst [hunk 7]
```diff
@@ -181,6 +181,12 @@ then your validator is already registered as a service and :doc:`tagged </servic
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
Comment (HeahDude) [inline validation/custom_constraint.rst]: Please add a `.. versionadded:: 5.1` directive

**Review by HeahDude — COMMENTED** [review]
Thanks for that cool feature :). Some minor comments about the naming and CS

File: reference/constraints/Compound.rst [hunk 8]
```diff
@@ -0,0 +1,110 @@
+Compound
+========
+
```
Comment (OskarStark) [inline reference/constraints/Compound.rst]: Please add a versionadded directive

File: reference/constraints/Compound.rst (hunk 4 above)
Comment (ogizanagi) [reply reference/constraints/Compound.rst]: We're using `AcmeAssert` for the whole `App\Validator\Constraints` namespace in https://symfony.com/doc/current/validation/custom_constraint.html#using-the-new-validator, maybe we should use it here too for consistency.

(but I personally prefer no alias at all, as there is no ambiguity to me)

**Comment by ogizanagi** [comment]
Thank you @OskarStark & @HeahDude for your reviews. I've applied your suggestions.

File: reference/constraints/Compound.rst [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+.. versionadded:: 5.1
+
+    The :class:`Symfony\\Component\\Validator\\Constraints\\Compound` class was
+    introduced in Symfony 5.1.
+
+To the contrary to the other constraints, this constraint cannot be used on its own.
+Instead, it allows you to create your own set of reusable constraints, representing
+rules to use consistently across your application, by extending the constraint.
+
```
Comment (OskarStark) [inline reference/constraints/Compound.rst]: Please move the versionadded below the introduction and reword it like this:
```
.. versionadded:: 5.1

    The ``Compound`` constraint was introduced in Symfony 5.1.
```

File: reference/constraints/Compound.rst [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+Validator   :class:`Symfony\\Component\\Validator\\Constraints\\CompoundValidator`
+==========  ===================================================================
+
+Basic Usage
+-----------
+
+Suppose that you have different places where a user password must be validated,
+you can create your own named set or requirements to be reused consistently everywhere:
+
+.. configuration-block::
```
Comment (OskarStark) [inline reference/constraints/Compound.rst]: No need for the `configuration-block` directive

File: reference/constraints/Compound.rst [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+                <property name="password">
+                    <constraint name="App\Validator\Constraints\PasswordRequirements"/>
+                </property>
+            </class>
+        </constraint-mapping>
+
+    .. code-block:: php
+
+        // src/Entity/User.php
+        namespace App\DTO;
```
Comment (OskarStark) [inline reference/constraints/Compound.rst]:
```suggestion
        namespace App\User;
```

File: reference/constraints/Compound.rst [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+            <class name="App\User\RegisterUser">
+                <property name="password">
+                    <constraint name="App\Validator\Constraints\PasswordRequirements"/>
+                </property>
+            </class>
+        </constraint-mapping>
+
+    .. code-block:: php
+
+        // src/Entity/User.php
```
Comment (OskarStark) [inline reference/constraints/Compound.rst]:
```suggestion
        // src/User/RegisterUser.php
```

**Review by OskarStark — APPROVED** [review]

**Review by HeahDude — APPROVED** [review]

**Comment by HeahDude** [comment]
Thank you Maxime.

**Pull request merged**